### PR TITLE
Allowing onBeforePrint to return a Promise

### PR DESCRIPTION
**Pull request by aviklai**
#140

### CODE REVIEW

File: src/index.tsx [hunk 1]
```diff
@@ -44,6 +38,23 @@ export default class ReactToPrint extends React.Component<IReactToPrintProps> {
...
+
+    triggerPrint = (target) => {
+        const { onBeforePrint, onAfterPrint } = this.props;
+
+        if (onBeforePrint) {
+            const onBeforePrintOutput = onBeforePrint();
+            if (onBeforePrintOutput instanceof Promise) {
+                onBeforePrintOutput.then(r => {
+                    this.startPrint(target, onAfterPrint);
+                }).catch(e => {
```
Comment (MatthewHerbst) [inline src/index.tsx]: I don't think we should `catch` this failure. If we do this then if the user chains a `.catch` it won't get fired.

File: src/index.tsx [hunk 2]
```diff
@@ -44,6 +38,23 @@ export default class ReactToPrint extends React.Component<IReactToPrintProps> {
...
         }, 500);
+    }
+
+    triggerPrint = (target) => {
+        const { onBeforePrint, onAfterPrint } = this.props;
+
+        if (onBeforePrint) {
+            const onBeforePrintOutput = onBeforePrint();
+            if (onBeforePrintOutput instanceof Promise) {
+                onBeforePrintOutput.then(r => {
```
Comment (MatthewHerbst) [inline src/index.tsx]: Nit: `r` isn't used and so doesn't need to be defined, should just be `()`

File: src/index.tsx [hunk 3]
```diff
@@ -14,7 +14,7 @@ export interface IReactToPrintProps {
     /** Copy styles over into print window. default: true */
     copyStyles?: boolean;
     /** Callback function to trigger before print */
-    onBeforePrint?: () => void;
+    onBeforePrint?: () => void | Promise<any>;
```
Comment (MatthewHerbst) [inline src/index.tsx:17]: Please also update the README

File: src/index.tsx [hunk 4]
```diff
@@ -44,6 +38,23 @@ export default class ReactToPrint extends React.Component<IReactToPrintProps> {
...
+
+        if (onBeforePrint) {
+            const onBeforePrintOutput = onBeforePrint();
+            if (onBeforePrintOutput instanceof Promise) {
+                onBeforePrintOutput.then(r => {
+                    this.startPrint(target, onAfterPrint);
+                }).catch(e => {
+                    console.error(`onBeforePrint was not resolved. error: ${e.toString()}`);
+                });
+            }else{
```
Comment (MatthewHerbst) [inline src/index.tsx]: Nit

```suggestion
            } else {                
```

**Review by MatthewHerbst — CHANGES_REQUESTED** [review]
Thanks for jumping on this! Just a few minor things and then I think we'll be good to go (we'll have to release this under a breaking change but I think that's ok)

File: src/index.tsx (hunk 3 above)
Comment (aviklai) [reply src/index.tsx:17]: @MatthewHerbst Done. Please tell me if the explanation is good enough.

**Comment by aviklai** [comment]
@MatthewHerbst Why is this a breaking change? we still allow the onBeforePrint function to return void.

File: src/index.tsx (hunk 1 above)
Comment (aviklai) [reply src/index.tsx]: Done.

File: src/index.tsx (hunk 2 above)
Comment (aviklai) [reply src/index.tsx]: Done.

File: src/index.tsx (hunk 4 above)
Comment (aviklai) [reply src/index.tsx]: Done.

**Comment by MatthewHerbst** [comment]
This line:

```js
onBeforePrintOutput instanceof Promise
```

will break if `Promise` does not exist. Specifically, if you run that on a browser (I tested on IE10 via Browserstack) without `Promise`, you will get the error:

```js
var x = {}
x instanceof Promise
// Error: "'Promise' is undefined"
```

So the implementation as it would still require users to polyfill `Promise` for environments it doesn't exist in, which is not currently a library requirement, which is why it would require a breaking change.

Maybe there is another way to check if something is a `Promise`?

**Comment by aviklai** [comment]
Apparently, there is no standardized way:
https://stackoverflow.com/questions/27746304/how-do-i-tell-if-an-object-is-a-promise

This is the es5 answer that we can use (and graphql-js use):
https://github.com/graphql/graphql-js/blob/master/src/jsutils/isPromise.js

So it would look like this:
`if (onBeforePrintOutput && typeof onBeforePrintOutput.then === 'function')`

What do you think?

**Comment by MatthewHerbst** [comment]
I'm ok with that. It's not 100% perfect (as some commenters in the accepted answer of that SO post point out) but I think it's good enough

**Comment by aviklai** [comment]
@MatthewHerbst Done - changed the way to check if it's a promise.

**Comment by MatthewHerbst** [comment]
Awesome! One last request: can you please squash the commits together?

**Comment by aviklai** [comment]
@MatthewHerbst Done.

**Review by MatthewHerbst — APPROVED** [review]
Awesome, thanks! I'll get a new version with this published in a little bit

**Comment by MatthewHerbst** [comment]
Published as v2.2.0. Thanks again!

**Pull request merged**